fedspendingtransparency/usaspending-api
# Updating README and local config

**Description:**
Update the README and other aspects of the `usaspending-api` codebase to support local development.

**Technical details:**
Below is a brief list of changes going file by file in order to give a high-level reasoning of "why".

* `.env.template`
   * The wrong database name was used for the Broker DB
   * Updated the rest of the DB string to create the Broker DB for tests on the USAspending DB container
* `.gitignore`
   * Change to make sure we are ignoring nested directories as well
* `CONTRIBUTING.md`
   * Update reference to Travis CI to be GitHub Actions
* `Dockerfile`
   * The `es_configure` job relies on `curl` to query the ES indexes
   * Without `curl` the `elasticsearch_indexer` job fails to create the templates for the ES indexes and thus creates many mis-typed fields on the indexes
* `Dockerfile.spark`
   * Similar changes that were made to the `Dockerfile` to change the base image used
   * Also had to update the version of OpenJDK; use AWS Corretto JDK 8 as it is the same version used by Databricks and this is a LTS version maintained by AWS
   * Some new installs were needed to support the different image
   * Old installs were left in place to avoid breaking things
* `Dockerfile.testing`
   * Update to the OpenJDK version similar to `Dockerfile.spark`
   * Remove the logic that changed the working directory in favor of having the volume named `/usaspending-api` instead of `/dockermount`
* Makefile
   * Update references of `docker-compose` to instead use `docker compose` (read more [here](https://docs.docker.com/compose/releases/migrate/#what-are-the-differences-between-compose-v1-and-compose-v2))
   * Added back reference to spark-hive in Makefile; I incorrectly removed this in the Python upgrade
* `README.md`
   * Some cleanup
   * Update all shell commands to support copy and paste as-is
   * Made changes to reflect the steps that I took when setting everything up
* `docker-compose.yaml`
   * Removed the `version` as that is deprecated
   * Updated the `usaspending-test` volume to be `/usaspending-api` in the container to support hot reload when testing
   * Change the folder create for bulk downloads on Minio; Minio folders cannot have an underscore in the name
* `pyproject.toml`
   * Add the current directory as pythonpath to support tests having issues to run
* `spark_helpers.py`
   * Added a check for the "SparkSession" attribute on the JVM; this projects against some cases where during cleanup the SparkSession was already closed / removed by another parallel test build
* `local.py`
   * Minio doesn't support bucket names with underscore
* `conftest.py`
   * The test suite handles creation of a Broker DB for tests via a docker run command. The network was proving to be a problem when attempting to run this via the docker containers exclusively. As a workaround, when using the test container, it will define a "PROJECT_SRC_PATH" env that is used to assign the created Broker containers to the same network, allowing reference by name.
* `conftest_spark.py`
   * Increase the memory allocated to Spark to support the test cases

**Requirements for PR merge:**

1. [x] Unit & integration tests updated
2. [x] API documentation updated
3. [x] Necessary PR reviewers:
4. [x] Matview impact assessment completed
5. [x] Frontend impact assessment completed
6. [x] Data validation completed
7. N/A Appropriate Operations ticket(s) created
8. N/A Jira Ticket

**Area for explaining above N/A when needed:**
```
```
